File: 2-chunked-upload-with-progress/backend/src/modules/upload/upload.service.ts
import {
    Injectable,
    Logger,
    NotFoundException,
    BadRequestException,
} from "@nestjs/common"
import {
    ConfigService,
} from "@nestjs/config"
import {
    createReadStream,
    createWriteStream,
    promises as fs,
} from "node:fs"
import {
    join,
} from "node:path"
import {
    createHash,
    randomUUID,
} from "node:crypto"
import {
    UPLOAD_CONFIG_TOKEN,
    type UploadConfigShape,
} from "../../config"
import type {
    UploadSession,
    InitSessionResponse,
    SessionStatusResponse,
    FinalizeResponse,
} from "./upload.types"

/**
 * UploadService — quản lý chunked upload session in-memory + ghi/đọc chunk file xuống disk.
 * (EN: UploadService — manages chunked upload sessions in-memory plus writes/reads chunk files to disk.)
 */
@Injectable()
export class UploadService {
    private readonly logger = new Logger(UploadService.name)
    private readonly cfg: UploadConfigShape
    private readonly sessions = new Map<string, UploadSession>()

    public constructor(configService: ConfigService) {
        this.cfg = configService.getOrThrow<UploadConfigShape>(UPLOAD_CONFIG_TOKEN)
    }

    /**
     * Tạo session mới — totalChunks = ceil(size / chunkSize), persist tmp folder ngay.
     * (EN: Create a new session — totalChunks = ceil(size / chunkSize), persist the tmp folder immediately.)
     */
    public async initSession(
        filename: string,
        size: number,
        chunkSize?: number,
    ): Promise<InitSessionResponse> {
        if (size <= 0 || size > this.cfg.maxFileBytes) {
            throw new BadRequestException(`Invalid size: must be 1..${this.cfg.maxFileBytes}`)
        }
        const effectiveChunkSize = chunkSize && chunkSize > 0 ? chunkSize : this.cfg.chunkSizeBytes
        const totalChunks = Math.ceil(size / effectiveChunkSize)
        const id = randomUUID()
        const session: UploadSession = {
            id,
            filename,
            size,
            chunkSize: effectiveChunkSize,
            totalChunks,
            receivedChunks: [],
            createdAt: Date.now(),
            finalized: false,
        }
        this.sessions.set(id, session)
        await fs.mkdir(join(this.cfg.tmpDir, id), { recursive: true })
        this.logger.log(`init session ${id} total=${totalChunks}`)
        return { sessionId: id, totalChunks, chunkSize: effectiveChunkSize }
    }

    /**
     * Ghi 1 chunk vào `<tmp>/<id>/<n>.part`. Idempotent — gọi lại cùng index ghi đè.
     * (EN: Write a single chunk to `<tmp>/<id>/<n>.part`. Idempotent — recalling the same index overwrites.)
     */
    public async writeChunk(id: string, index: number, buffer: Buffer): Promise<void> {
        const session = this.requireSession(id)
        if (!Number.isInteger(index) || index < 0 || index >= session.totalChunks) {
            throw new BadRequestException(`Chunk index ${index} out of range [0, ${session.totalChunks})`)
        }
        const path = join(this.cfg.tmpDir, id, `${index}.part`)
        await fs.writeFile(path, buffer)
        if (!session.receivedChunks.includes(index)) {
            session.receivedChunks.push(index)
            session.receivedChunks.sort((a, b) => a - b)
        }
        this.logger.log(`received chunk ${index}/${session.totalChunks} (${buffer.length}B)`)
    }

    /**
     * Trả về status + bitmap để client biết chunk nào còn thiếu khi resume.
     * (EN: Return status + bitmap so the client knows which chunks are still missing on resume.)
     */
    public getStatus(id: string): SessionStatusResponse {
        const session = this.requireSession(id)
        const received = [...session.receivedChunks]
        const missing: number[] = []
        for (let i = 0; i < session.totalChunks; i++) {
            if (!session.receivedChunks.includes(i)) missing.push(i)
        }
        return {
            sessionId: session.id,
            totalChunks: session.totalChunks,
            chunkSize: session.chunkSize,
            received,
            missing,
            finalized: session.finalized,
        }
    }

    /**
     * Merge mọi chunk theo thứ tự + tính SHA-256 cùng pass + cleanup tmp folder.
     * (EN: Merge every chunk in order, compute SHA-256 in the same pass, then clean up the tmp folder.)
     */
    public async finalize(id: string): Promise<FinalizeResponse> {
        const session = this.requireSession(id)
        // Đảm bảo đủ chunk trước khi merge.
        // (EN: Ensure all chunks present before merging.)
        if (session.receivedChunks.length !== session.totalChunks) {
            throw new BadRequestException(
                `Missing chunks: ${session.totalChunks - session.receivedChunks.length}`,
            )
        }
        await fs.mkdir(this.cfg.finalDir, { recursive: true })
        const finalPath = join(this.cfg.finalDir, `${id}-${session.filename}`)
        const out = createWriteStream(finalPath)
        const hash = createHash("sha256")
        let size = 0
        for (let i = 0; i < session.totalChunks; i++) {
            const rs = createReadStream(join(this.cfg.tmpDir, id, `${i}.part`))
            // Stream-concat theo đúng thứ tự + tính SHA-256 đồng thời.
            // (EN: Stream-concatenate in order and compute SHA-256 in the same pass.)
            await new Promise<void>((resolve, reject) => {
                rs.on("data", (chunk: Buffer | string) => {
                    const buf = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk)
                    hash.update(buf)
                    size += buf.length
                })
                rs.on("end", () => resolve())
                rs.on("error", reject)
                rs.pipe(out, { end: false })
            })
        }
        await new Promise<void>((resolve, reject) => {
            out.end(() => resolve())
            out.on("error", reject)
        })
        await fs.rm(join(this.cfg.tmpDir, id), { recursive: true, force: true })
        session.finalized = true
        const sha256 = hash.digest("hex")
        this.logger.log(`finalized -> ${finalPath}`)
        return { filename: session.filename, size, sha256, path: finalPath }
    }

    /**
     * Throw 404 nếu session không tồn tại — wrap cho mọi handler dùng id.
     * (EN: Throw 404 when the session does not exist — wrapper for every handler using id.)
     */
    private requireSession(id: string): UploadSession {
        const session = this.sessions.get(id)
        if (!session) throw new NotFoundException(`Upload session ${id} not found`)
        return session
    }
}

File: 2-chunked-upload-with-progress/backend/2-csharp/Services/UploadService.cs
using System.Collections.Concurrent;
using System.Security.Cryptography;
using ChunkedUpload.Models;

namespace ChunkedUpload.Services;

/// <summary>
/// Manages chunked upload sessions in memory and writes/reads chunk files to disk.
/// Sessions are stored in a ConcurrentDictionary (equivalent to TypeScript Map).
/// </summary>
public sealed class UploadService
{
    private readonly ILogger<UploadService> _logger;

    // Env-configured paths and limits
    private readonly string _tmpDir;
    private readonly string _finalDir;
    private readonly long _chunkSizeBytes;
    private readonly long _maxFileBytes;

    // In-memory session store — thread-safe equivalent of TypeScript Map<string, UploadSession>
    private readonly ConcurrentDictionary<string, UploadSession> _sessions = new();

    // Per-session lock objects so concurrent PATCH requests for the same session are serialized
    private readonly ConcurrentDictionary<string, SemaphoreSlim> _locks = new();

    public UploadService(ILogger<UploadService> logger)
    {
        _logger = logger;

        _tmpDir = Environment.GetEnvironmentVariable("UPLOAD_TMP_DIR") ?? "tmp";
        _finalDir = Environment.GetEnvironmentVariable("UPLOAD_FINAL_DIR") ?? "uploads";
        _chunkSizeBytes = ParseEnvLong("UPLOAD_CHUNK_SIZE_BYTES", 5L * 1024 * 1024);
        _maxFileBytes = ParseEnvLong("UPLOAD_MAX_FILE_BYTES", 1024L * 1024 * 1024);
    }

    /// <summary>
    /// Create a new session. totalChunks = ceil(size / chunkSize).
    /// Immediately creates the tmp directory for this session.
    /// </summary>
    public async Task<InitSessionResponse> InitSessionAsync(string filename, long size, long? chunkSize)
    {
        if (size <= 0 || size > _maxFileBytes)
            throw new ArgumentOutOfRangeException(nameof(size), $"Invalid size: must be 1..{_maxFileBytes}");

        var effectiveChunkSize = chunkSize is > 0 ? chunkSize.Value : _chunkSizeBytes;
        var totalChunks = (int)Math.Ceiling((double)size / effectiveChunkSize);
        var id = Guid.NewGuid().ToString();

        var session = new UploadSession
        {
            Id = id,
            Filename = filename,
            Size = size,
            ChunkSize = effectiveChunkSize,
            TotalChunks = totalChunks,
            CreatedAt = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
            Finalized = false,
        };

        _sessions[id] = session;
        _locks[id] = new SemaphoreSlim(1, 1);

        // Create the tmp directory eagerly so PATCH can write immediately
        Directory.CreateDirectory(Path.Combine(_tmpDir, id));

        _logger.LogInformation("init session {Id} total={Total}", id, totalChunks);
        return await Task.FromResult(new InitSessionResponse(id, totalChunks, effectiveChunkSize));
    }

    /// <summary>
    /// Write a single chunk to tmp/&lt;id&gt;/&lt;index&gt;.part.
    /// Idempotent — re-sending the same index overwrites the existing part file.
    /// </summary>
    public async Task WriteChunkAsync(string id, int index, byte[] buffer)
    {
        var session = RequireSession(id);

        if (index < 0 || index >= session.TotalChunks)
            throw new ArgumentOutOfRangeException(nameof(index),
                $"Chunk index {index} out of range [0, {session.TotalChunks})");

        var partPath = Path.Combine(_tmpDir, id, $"{index}.part");
        await File.WriteAllBytesAsync(partPath, buffer);

        // Serialize ReceivedChunks mutation to avoid race conditions between concurrent PATCHes
        var sem = _locks.GetOrAdd(id, _ => new SemaphoreSlim(1, 1));
        await sem.WaitAsync();
        try
        {
            if (!session.ReceivedChunks.Contains(index))
            {
                session.ReceivedChunks.Add(index);
                session.ReceivedChunks.Sort();
            }
        }
        finally
        {
            sem.Release();
        }

        _logger.LogInformation("received chunk {Index}/{Total} ({Bytes}B)", index, session.TotalChunks, buffer.Length);
    }

    /// <summary>
    /// Return session status + bitmap so the client knows which chunks are still missing on resume.
    /// </summary>
    public SessionStatusResponse GetStatus(string id)
    {
        var session = RequireSession(id);

        List<int> received;
        List<int> missing = new();

        // Snapshot under lock to get a consistent view
        var sem = _locks.GetOrAdd(id, _ => new SemaphoreSlim(1, 1));
        sem.Wait();
        try
        {
            received = new List<int>(session.ReceivedChunks);
        }
        finally
        {
            sem.Release();
        }

        for (var i = 0; i < session.TotalChunks; i++)
        {
            if (!received.Contains(i)) missing.Add(i);
        }

        return new SessionStatusResponse(
            session.Id,
            session.TotalChunks,
            session.ChunkSize,
            received,
            missing,
            session.Finalized
        );
    }

    /// <summary>
    /// Merge every chunk in order, compute SHA-256 in the same pass, then clean up tmp folder.
    /// Returns the merged file metadata matching the TypeScript FinalizeResponse shape.
    /// </summary>
    public async Task<FinalizeResponse> FinalizeAsync(string id)
    {
        var session = RequireSession(id);

        if (session.ReceivedChunks.Count != session.TotalChunks)
            throw new InvalidOperationException(
                $"Missing chunks: {session.TotalChunks - session.ReceivedChunks.Count}");

        Directory.CreateDirectory(_finalDir);

        var finalPath = Path.Combine(_finalDir, $"{id}-{session.Filename}");

        using var sha256 = SHA256.Create();
        long mergedSize = 0;

        // Stream-concatenate chunks in order and compute SHA-256 in the same pass
        await using (var outStream = new FileStream(finalPath, FileMode.Create, FileAccess.Write, FileShare.None))
        {
            for (var i = 0; i < session.TotalChunks; i++)
            {
                var partPath = Path.Combine(_tmpDir, id, $"{i}.part");
                var partBytes = await File.ReadAllBytesAsync(partPath);

                sha256.TransformBlock(partBytes, 0, partBytes.Length, null, 0);
                mergedSize += partBytes.Length;

                await outStream.WriteAsync(partBytes);
            }

            sha256.TransformFinalBlock(Array.Empty<byte>(), 0, 0);
        }

        // Clean up the tmp directory for this session
        var tmpSessionDir = Path.Combine(_tmpDir, id);
        if (Directory.Exists(tmpSessionDir))
            Directory.Delete(tmpSessionDir, recursive: true);

        session.Finalized = true;

        var hexHash = Convert.ToHexString(sha256.Hash!).ToLowerInvariant();
        _logger.LogInformation("finalized -> {Path}", finalPath);

        return new FinalizeResponse(session.Filename, mergedSize, hexHash, finalPath);
    }

    // ---------------------------------------------------------------------------
    // Helpers
    // ---------------------------------------------------------------------------

    /// <summary>
    /// Throw 404-equivalent when the session does not exist.
    /// The controller maps KeyNotFoundException to 404.
    /// </summary>
    private UploadSession RequireSession(string id)
    {
        if (!_sessions.TryGetValue(id, out var session))
            throw new KeyNotFoundException($"Upload session {id} not found");
        return session;
    }

    /// <summary>Parse an environment variable as long; fall back to defaultValue on failure.</summary>
    private static long ParseEnvLong(string name, long defaultValue)
    {
        var raw = Environment.GetEnvironmentVariable(name);
        return long.TryParse(raw, out var v) && v > 0 ? v : defaultValue;
    }
}
